src/coreclr/src/jit/morph.cpp

```diff
@@ -10830,6 +10830,176 @@ GenTree* Compiler::fgMorphFieldAssignToSIMDIntrinsicSet(GenTree* tree)
 
 #endif // FEATURE_SIMD
 
+//------------------------------------------------------------------------
+// fgMorphCommutative: Fold constants in a tree with commutative operators (|, ^, &, +)
+//
+// Arguments:
+//    tree - This node will be checked for patterns where constants can be folded.
+//
+// Return Value:
+//    A GenTree* which points to the new tree with folded constants.
+//
+
+GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+{
+    assert(tree->OperIs(GT_OR, GT_XOR, GT_AND, GT_ADD));
+
+    GenTree*   op1  = tree->gtGetOp1();
+    GenTree*   op2  = tree->gtGetOp2();
+    genTreeOps oper = tree->OperGet();
+
+    // Fold "((x <op> icon1) <op> (y <op> icon2))" to "((x <op> y) <op> (icon1 <op> icon2))"
+    if (op1->OperIs(op2->OperGet()) && op1->OperIs(oper) && !gtIsActiveCSE_Candidate(op2) &&
+        op1->AsOp()->gtGetOp2()->IsCnsIntOrI() && op2->AsOp()->gtGetOp2()->IsCnsIntOrI())
+    {
+        // Don't create a byref pointer that may point outside of the ref object.
+        // If a GC happens, the byref won't get updated. This can happen if one
+        // of the int components is negative. It also requires the address generation
+        // be in a fully-interruptible code region.
+        if (!varTypeIsGC(op1->AsOp()->gtGetOp1()->TypeGet()) && !varTypeIsGC(op2->AsOp()->gtGetOp1()->TypeGet()))
```

Comment:
here you check `varTypeIsGC(x)`, `varTypeIsGC(y)`, `gtIsActiveCSE_Candidate(op2)`.

Reply:
Added also `gtIsActiveCSE_Candidate(op1)` here